### PR TITLE
Fix nop -p corrupting stack, fix nop, turn nop -p into stub command

### PR DESCRIPTION
**Pull request by Grazfather**
* Fixes bug in issue in #117
* Fixes another issue where nop just basically didn't work
* Splits `nop -p` into its own `stub` command.

### CODE REVIEW

File: docs/commands/stub.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+## Command stup ##
```
Comment (Grazfather) [inline docs/commands/stub.md]: typo

File: docs/commands/stub.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+```
+gef➤ stub [-h] [-r RETVAL] [LOCATION]
+```
+
+`LOCATION` indicates the address of the function to bypass. If not
+specified, gef will consider the instruction at the program counter to be the
+start of the function.
+
+If `-r <return value>` is provided, gef will set the return value to the
```
Comment (Grazfather) [inline docs/commands/stub.md]: s/return value/RETVAL

File: gef.py [hunk 3]
```diff
@@ -2436,57 +2436,36 @@ def stop(self):
...
 
         m = "All calls to '{:s}' will be skipped".format(self.func)
         if self.retval is not None:
-            m += " (with return value as {:#x})".format(self.retval)
+            m += " (with return value set to {:#x})".format(self.retval)
         info(m)
         return
 
     def stop(self):
         retaddr = gdb.selected_frame().older().pc()
```
Comment (Grazfather) [inline gef.py]: not needed anymore

File: gef.py [hunk 4]
```diff
@@ -4078,75 +4056,102 @@ def get_insn_size(self, addr):
...
-        for o,a in opts:
-            if   o == "-r":
-                retval = long(a, 16)
-            elif o == "-p":
-                perm_mode = True
+        opts, args = getopt.getopt(argv, "b:h")
+        num_bytes = 0
+        for o, a in opts:
+            if o == "-b":
+                num_bytes = long(a, 0)
```
Comment (Grazfather) [inline gef.py]: changing from 16 to 0 allows the user to provide hex OR decimal (but hex must be precede with `0x`)

File: gef.py [hunk 5]
```diff
@@ -4078,75 +4056,102 @@ def get_insn_size(self, addr):
...
-        if len(nops) < size:
-            warn("Adjusting NOPs to size {:d}".format(size))
-            while len(nops) < size:
-                nops += current_arch.nop_insn
+        while len(nops) < size:
+            nops += current_arch.nop_insn
 
         if len(nops) != size:
-            err("Cannot patch instruction at {:#x} (unexpected NOP length)".format(loc))
+            err("Cannot patch instruction at {:#x} (nop instruction does not even fit in requested size)"
```
Comment (Grazfather) [inline gef.py]: s/even/evenly

File: gef.py [hunk 6]
```diff
@@ -4078,75 +4056,102 @@ def get_insn_size(self, addr):
...
-            reg = current_arch.return_register
-            addr = "*{}".format(format_address(loc))
-            SetRegisterBreakpoint(addr, reg, retval)
         return
 
 
-    def permanent_patch(self, loc, retval):
-        PatchBreakpoint(loc, retval)
+class StubCommand(GenericCommand):
+    """Stub out the specified function"""
```
Comment (Grazfather) [inline gef.py]: period at end of docstring

File: gef.py [hunk 7]
```diff
@@ -4078,75 +4056,102 @@ def get_insn_size(self, addr):
...
+
+    _cmdline_ = "stub"
+    _syntax_  = "{:s} [-r RETVAL] [-h] [LOCATION]".format(_cmdline_)
+
+
```
Comment (Grazfather) [inline gef.py]: extra line

File: gef.py [hunk 8]
```diff
@@ -4078,75 +4056,102 @@ def get_insn_size(self, addr):
...
+    def __init__(self):
+        super(StubCommand, self).__init__(complete=gdb.COMPLETE_LOCATION, prefix=False)
+        return
+
+
```
Comment (Grazfather) [inline gef.py]: extra line

File: gef.py [hunk 9]
```diff
@@ -4078,75 +4056,102 @@ def get_insn_size(self, addr):
...
+                retval = long(a, 0)
+            elif o == "-h":
+                self.help()
+                return
+
+        if not args:
+            loc = "*{:#x}".format(current_arch.pc)
+        else:
+            loc = args[0]
+        print(loc) # DELETEME
```
Comment (Grazfather) [inline gef.py]: oops

File: gef.py [hunk 10]
```diff
@@ -4078,75 +4056,102 @@ def get_insn_size(self, addr):
...
+
+        self.stub_out(loc, retval)
+        return
+
```
Comment (Grazfather) [inline gef.py]: extra line

File: gef.py [hunk 11]
```diff
@@ -4078,75 +4056,102 @@ def get_insn_size(self, addr):
...
+
+    def help(self):
+        m = self._syntax_
```
Comment (Grazfather) [inline gef.py]: use join idiom

**Comment by Grazfather** [comment]
LGTM.

File: gef.py [hunk 12]
```diff
@@ -3506,7 +3483,7 @@ def do_invoke(self, argv):
             else:
                 new_flags = old_flag ^ (1<<off)
 
-            gdb.execute("set ({:s}) = {:x}".format(current_arch.flag_register, new_flags))
+            gdb.execute("set ({:s}) = {:#x}".format(current_arch.flag_register, new_flags))
```
Comment (Grazfather) [inline gef.py]: We had broken flags :(

File: docs/commands/stub.md [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+### Example ###
+
+Patching `fork()` calls:
+
+* Without stub:
+![fork execution](http://i.imgur.com/TjnTDot.png)
+
+* With stub:
+![stubbed fork](http://i.imgur.com/CllTnRH.png)
```
Comment (Grazfather) [inline docs/commands/stub.md]: This are uploaded to my imgur account. @hugsy you might want to re-upload to your own and then update the urls.

Comment (hugsy) [inline docs/commands/stub.md]: I don't have an imgur account, I use this : https://github.com/jomo/imgur-screenshot

Comment (Grazfather) [inline docs/commands/stub.md]: That's cool. I already pretty much use something like that (skitch). We should be good, then.

**Pull request merged**